WIP - blockchain catchup

yusefnapora opened the pull request:
Adds a `RocksDatastore` that writes to a local rocksdb
- so far doesn't add any setup or provisioning scripts to ensure that rocksdb is installed, so this will fail if it can't find the rocksdb lib (works with `brew install rocksdb` on a mac)

Adds a read-through `BlockCache` that fetches blocks from a remote datastore and saves them to rocks.

Adds a naive `BlockchainFollower` class that runs a worker thread to chase `chain` references back to the beginning of the blockchain.  At the moment, it will always follow the entire chain back to the beginning (via the local cache, so at least it's not spamming the network).

todo:
- [x] provisioning setup - need to build / install rocksdb
- [ ] update README with rocksdb dependency, install instructions
- [ ] allow configuring how far back to follow the chain
- [x] notify when catchup is complete / expose completion state
- [ ] hook up to the transactor client
  - [x] incoming blocks from the `journal_stream` should be piped in to update the local block cache
  - [ ] we should expose a stream of canonical references, so we can iterate over the set of all known canonicals from the blockchain.

If we want to have the "canonical stream" include all historical entries and future entries in a single stream, we need to queue up incoming entries until catchup is complete, then play those back onto the stream.

yusefnapora committed:
basic blockchain catchup impl

yusefnapora committed:
try to cbor encode dictionaries before storing in rocksdb

yusefnapora committed:
wip - replay events from BlockchainFollower

parkan commented:
Hmm not sure why this build of rocksdb is failing, I install gcc-4.8 already. Maybe it needs some cmldline flags.

yusefnapora commented:
oh, huh. I didn't realize it was actually trying to build rocks; I though it was building its own c extension that required rocks to already be installed...  but maybe it tries to build if if its missing.

`cc1plus: error: unrecognized command line option ‘-std=c++11’`

maybe need to update g++?

bigs commented on mediachain/datastore/rocks.py:
```diff
@@ -0,0 +1,51 @@
+
+class RocksDatastore(object):
+    def __init__(self, db_path):
+        opts = rocksdb.Options(create_if_missing=True)
```
maybe splice in some `kwargs`

bigs commented on mediachain/datastore/rocks.py:
```diff
@@ -0,0 +1,51 @@
+        except ValueError:
+            return byte_string
+
+    def put(self, data_object, ref=None):
```
having ref logic doesn't really make sense in this case

yusefnapora replied:
you mean it should just be `key` instead of ref, and not try to compute the hash if no key is given?

parkan replied:
I think @bigs means that multihash calculation shouldn't be present above the interface, because our target iface (IPFS) encapsulates it?

bigs replied:
yeah like it doesn't make sense to "put" to a specific key b/c it's content-addressed

yusefnapora replied:
That's true; I think I added the ref key as a paranoid guard against the transactor changing its multihash calculation or something, but that would break so many other things that it's silly to change this interface for that.  I'll fix that up when I circle back to this branch :)

parkan commented:
OK I tried to fix the build, it gets farther now but still dies on 

```
    rocksdb/_rocksdb.cpp:350:27: fatal error: rocksdb/slice.h: No such file or directory
     #include "rocksdb/slice.h"
                               ^
    compilation terminated.
    error: command 'gcc-4.8' failed with exit status 1
```

parkan commented:
Yeah ok seems like you need to install rocksdb headers etc https://github.com/stephan-hof/pyrocksdb/issues/1

yusefnapora committed:
add RocksDatastore implementation

yusefnapora committed:
Need g++

yusefnapora committed:
add script to build rocksdb and install pyrocksdb package, run on travis

yusefnapora committed:
fix yaml error in travis config

yusefnapora committed:
use ubuntu trusty for travis

yusefnapora committed:
[travis] add rocksdb extension to pylint whitelist

yusefnapora committed:
[travis] quotes around pip install pylint cmd

yusefnapora committed:
[travis] try ignoring rocksdb in pylint

yusefnapora committed:
rm RocksDatastore, use rocks directly in BlockCache

yusefnapora committed:
add timeout to event replay method

yusefnapora committed:
merge catchup events with incoming event stream

parkan commented:
Needs to be rebased on master, what else is missing?

yusefnapora commented:
there's a couple rough edges here still.  the stream can only be iterated once, which might be fine; a new call to `client.journal_stream()` will make a new `BlockchainFollower` that can be iterated again.  

Also, if you disable the timeout for the RPC journal stream like I'm doing right now, you have to call `.cancel()`, or else gRPC will keep the program alive forever.  I tried adding a `__del__`, to call cancel when it gets GC'd, but I must have a reference cycle in there that's preventing `__del__` from being called.  

So I need to either bring back the timeouts, fix `__del__`, or make it really explicit that you have to call `.cancel()` when you're done.

parkan commented:
Well, it kind of makes sense that it'll stay alive if it's in follow mode. We should put the cancel into an interrupt handler, though

yusefnapora committed:
use contextmanager to support `with/as` blocks for journal stream

yusefnapora commented:
okay, I figured out how to make the `BlockchainFollower` class (and the non-catchup journal_stream) work with python's `contextmanager` protocol, so you can do

```python
with client.journal_stream() as stream:
  for event in stream:
    # do something with event
```

and it will call `.cancel` when it goes out of the `with` block scope.  That seems like it'll do; we just need to be careful to use the `with` idiom, or remember to call `.cancel` on the returned stream.

yusefnapora commented:
ugh.... when I have the infinite timeout, it seems to just swallow all `KeyboardInterrupt` exceptions.... so frustrating.  I really hate the grpc native / python exception handling mess.

yusefnapora commented:
okay, I think I need to rethink this a bit...

Treating this stream like an infinite generator is problematic, since it can block for an indefinite time between events, and apparently it swallows all Exceptions while it's blocked 😡.

So maybe it's better for the `journal_stream` method to accept a callback that it pushes events to.  And it can just return an object with a `cancel` method so you can control the lifecycle of the stream.

yusefnapora commented:
I kind of hate the callback approach for different reasons though...  I'm going to see what happens if I read from the stream on a background thread.  Hopefully it won't interfere with exception handling then, and we can just make sure we always consume the stream off the main thread, which we have to do anyway if we want the stream to be infinite.

parkan commented:
Hmm does the GIL apply to C threads? It seems like it should still yield GIL every 100 ticks or w/e even if it's blocked on I/O, right?

yusefnapora commented:
I'm not sure... cython has a `nogil` context that is used in a few places in grpc, which is supposed to indicate that the code in that context can be run without the GIL.   So I think you have to explicitly opt-out of the GIL, and maybe gprc doesn't do so for streaming calls?  

idk.. so far all I know is that if you trap `KeyboardInterrupt` while waiting on a streaming rpc call, it will only be caught after the timeout expires.  So if the timeout is infinite, it never gets caught 😄.  That's when iterating over the stream on the main thread though (in the repl).  Need to finish testing the background consumer thread thing

parkan commented:
OK, I am holding my breath for background thread (signal handler only installed in main)

yusefnapora commented:
alright, it looks like if you consume the stream in a background thread, you can kill the process with Ctrl-C.  I need to clean up a few things, but I think this will work.

yusefnapora committed:
return a JournalStreamConsumer from client.journal_stream that reads on a bg thread

yusefnapora commented:
well, this still needs some work...  but it's better :)  now `client.journal_stream()` returns a `JournalStreamConsumer`, which reads the rpc events (and catchup events, if `catchup=True`) on a background thread.

The default behavior of the `JournalStreamConsumer` is to save events in a list, which you can iterate over by iterating over the consumer itself... But you can also set `consumer.on_event` to your own function to change the behavior.  That default behavior will get us into trouble when we start having millions of events... if we want the consumer to be iterable, we should use a queue instead of a list, so we don't eat up unbounded memory

At the moment, you have to call `.start()` on the consumer before it will start reading events; that should maybe be automatic.

Anyway, I'll try to wrap this up tomorrow

yusefnapora commented:
Okay, it turns out that the exception swallowing was not being caused by grpc's timeouts after all!  It was because I'm using the blocking form of `Queue.get`, to retrieve journal events from a replay queue inside of `BlockchainFollower`.  If I change to the non-blocking form with a timeout and sleep for a bit when the queue is empty, it will process `KeyboardInterrupt` exceptions during the sleep.

Apparently when the python docs say something is blocking, they mean business 😛 I assumed that a thrown exception would interrupt the blocking get, but I guess not.

So anyway, the stream consumer isn't necessary, I think.  Will pull it out and use the timeout + sleep approach.

yusefnapora committed:
rm JournalStreamConsumer, use non-blocking Queue.get

yusefnapora committed:
use BlockchainFollower for all journal streams, not just catchup

yusefnapora committed:
persist block cache to disk by default, use a shared cache for all clients

yusefnapora committed:
doc comment for journal_stream

yusefnapora committed:
bring back timeout for canonical_stream temporarily

- the canonical_stream method needs to close the underlying
  journal_stream, or gRPC will prevent the process from exiting

- since the canonical_stream method needs a rewrite anyway, this is
  a temporary hack to ensure the indexer cleans up after itself

yusefnapora commented:
Okay, I'm starting to be mostly happy with this...

The catchup takes quite a while on the first run, about a minute on the current devnet (which has a fairly small subset of the data).  The second run is nearly instant, since I started persisting the block cache between invocations.   It might be a good idea to somehow warn the user on the first run that we're bootstrapping the blockchain and it will take a while, etc. 

The other thing is that you have to cleanup the stream, either by calling `.cancel()` on it or using 

```python
with client.journal_stream as stream:
  for event in stream:
    # ...
```

If the stream is still open, grpc won't let the process exit :( 

I added a doc comment to that effect, but the `canonical_stream` needs to be updated so that it closes the stream.  I haven't done that yet because that method needs to be rewritten anyway, and I need to think about that implementation a bit more.  Until then, I brought back the default timeout for the `canonical_stream` method, so it will have the same behavior as before and the indexer (which uses `canonical_stream`) won't be left holding the connection open.

The other thing that kind of sucks is the rocksdb dependency; it's easy on OS X where we can `brew install rocksdb`, but we need to update the README and give instructions for installing on linux.

parkan commented:
2 Q's:

1. Do you think it's important to have this in original testnet release?
2. What, if any, is the overlap between functionality here and in indexer?

yusefnapora commented:
1. I'd like to include it, because without it the system will appear empty to any clients that haven't been tailing the journal stream from the beginning. 
We could cut it from the initial release and just say that all query / discovery has to go through the indexer... but if the indexer goes down at all there will be gaps, since it's currently only responding to incoming events as they happen.

2. I don't think there's much overlap really.  The indexer is using the `canonical_stream` now to tail the blockchain, so adding catchup support will just let it find records that were ingested before it started listening.  In theory the client could use this to do some things that are more in the indexer's domain, like making an L-Space sort of graph or something.  But there's nothing like that in there now.

yusefnapora commented:
by "the system will appear empty" I mean that you'd need to know the canonical IDs to query from out-of-band knowledge.  You could still fetch records

parkan commented:
It's actually not entirely clear to me how users can inspect the state of the system that's obtained through this tailing -- the CLI really only presents the `get` API op?

yusefnapora commented:
that's a good point 😄

Maybe lets postpone this after all, since it would be an internal / developer-facing feature.  Unless we want to add a `list_ids` command or something, but then the scope keeps creeping...

parkan commented:
yeah let's cut it ✂️

vyzo commented:
I think we need catchup functionality for the client, because we don't otherwise expose the current block with
an API.

parkan commented:
Aight let me bring this up to master

Commit:
Merge remote-tracking branch 'origin/master' into yn-catchup

 Conflicts (resolved in favor of yn-catchup):
	mediachain/transactor/client.py

parkan commented on .travis.yml:
```diff
@@ -1,18 +1,24 @@
+  - ./scripts/install_pyrocksdb.sh $HOME/src-dependencies 4.8 0.4
   - pip install -r requirements.txt
-  - pip install pylint>1.6.1
+  - pip install 'pylint>=1.6.1'
```
was this an issue?

yusefnapora replied:
I'm not sure... I was flailing around trying to get things working on travis when I did that.  although when I ran that command locally, I ended up with a file named `1.6.1` in my working dir 😄

parkan replied:
hrm you are right, it seemed to work but I guess that's because > 1.6.1 just became latest

parkan commented:
OK travis is out of machines rn or something, gonna merge
